### PR TITLE
Add Tests to Verify Empty User Groups Doesn't Allow Users to Spawn Notebooks(ODS-572)

### PR DESCRIPTION
**Pull request by prudhvi-165**
Signed-off-by: prudhvi <pnammi@redhat.com>

### CODE REVIEW

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/806/
Result: Passed

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/825/

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 1]
```diff
@@ -15,6 +15,16 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
     Verify Default Access Groups Settings
     Verify User Can Spawn A Notebook
 
+Verify Empty Group Doesnt Allow Users To Spawn Notebooks
+    [Documentation]   Verify that USer is unable to Access Jupyterhub after modifying Access Groups in rhods-groups-config
+    [Tags]    Sanity
+    ...       ODS-572
```
Comment (lugi0) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:21]: Missing `Tier1` as tag and typo in the documentation (`USer`)

**Review by lugi0 — CHANGES_REQUESTED** [review]

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 1 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated

**Review by lugi0 — DISMISSED** [review]
lgtm

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 2]
```diff
@@ -24,3 +35,9 @@ Verify User Can Spawn A Notebook
     Wait For RHODS Dashboard To Load
     Launch JupyterHub Spawner From Dashboard
     Spawn Notebook With Arguments    image=s2i-minimal-notebook    size=Default
+
+Update Access Groups In RHODS Groups Config
```
Comment (jgarciao) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:39]: In ODS.robot there is a keyword `Set Default Access Groups Settings` that does the same

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 3]
```diff
@@ -15,6 +15,17 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
     Verify Default Access Groups Settings
     Verify User Can Spawn A Notebook
 
+Verify Empty Group Doesnt Allow Users To Spawn Notebooks
+    [Documentation]   Verify that User is unable to Access Jupyterhub after modifying Access Groups in rhods-groups-config
+    [Tags]    Sanity
+    ...       Tier1
+    ...       ODS-572
+    Run  oc patch configmap rhods-groups-config -n redhat-ods-applications --type="json" -p='[{"op":"replace", "path":"/data/admin_groups", "value":""}]'
```
Comment (jgarciao) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:23]: I think lines 23-24 can be replaced by a call to the keyword `Apply Access Groups` Settings in ODS.robot

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot [hunk 4]
```diff
@@ -53,11 +53,12 @@ Verify Anaconda Commercial Edition Fails Activation When Key Is Invalid
 
 Verify User Is Able to Activate Anaconda Commercial Edition
   [Tags]  Tier2
-  ...     ODS-272  ODS-344  ODS-501  ODS-588  ODS-1082
+  ...     ODS-272  ODS-344  ODS-501  ODS-588  ODS-1082  ODS-462
```
Comment (jgarciao) [inline tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:56]: Is it possible that this change should be in another PR? It doesn't seem related

**Review by jgarciao — CHANGES_REQUESTED** [review]

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot (hunk 4 above)
Comment (prudhvi-165) [reply tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:56]: Yes, it is a different test. Since it is a very small change , can we put it in this test?

Comment (jgarciao) [reply tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:56]: ok. Please remember to update polarion once is merged

Comment (prudhvi-165) [reply tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:56]: Okay

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 5]
```diff
@@ -15,6 +15,17 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
     Verify Default Access Groups Settings
     Verify User Can Spawn A Notebook
 
+Verify Empty Group Doesnt Allow Users To Spawn Notebooks
+    [Documentation]   Verify that User is unable to Access Jupyterhub after modifying Access Groups in rhods-groups-config
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:19]: I think doc should specify that the user is unable to access if the groups are set as empty

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 6]
```diff
@@ -15,6 +15,17 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
...
+    [Tags]    Sanity
+    ...       Tier1
+    ...       ODS-572
+    Run  oc patch configmap rhods-groups-config -n redhat-ods-applications --type="json" -p='[{"op":"replace", "path":"/data/admin_groups", "value":""}]'
+    Run  oc patch configmap rhods-groups-config -n redhat-ods-applications --type="json" -p='[{"op":"replace", "path":"/data/allowed_groups", "value":""}]'
+    Run  oc rollout latest dc/jupyterhub -n redhat-ods-applications
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:25]: there is kw called `Rollout JupyterHub` which does the same

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot [hunk 7]
```diff
@@ -99,3 +100,8 @@ Anaconda Commercial Edition Suite Setup
   [Documentation]  Setup for ACE test suite
   Set Library Search Order  SeleniumLibrary
   RHOSi Setup
+
+Verify Anaconda In Kfdef
+    [Documentation]  Verifies if Anaconda is present in Kfdef
+    ${application_names}=  Run  oc get kfdef -n redhat-ods-applications -o json | jq ' .items[0] | select(.metadata.name="opendatahub") | [.spec.applications[].name]'
```
Comment (bdattoma) [inline tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:106]: could you pls use the openshiftlibrary instead of `Run oc ...` ?

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/894/

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot (hunk 7 above)
Comment (prudhvi-165) [reply tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 6 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 5 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 3 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 2 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated

**Comment by bdattoma** [comment]
@prudhvi-165 please rebase the branch and test it once again. Moreover, what this the expected error while spawning a jupyterlab server using empty groups?

**Comment by prudhvi-165** [comment]
> @prudhvi-165 please rebase the branch and test it once again. Moreover, what this the expected error while spawning a jupyterlab server using empty groups?

@bdattoma  It will not allow us to spawn notebook.

**Comment by prudhvi-165** [comment]
> @prudhvi-165 please rebase the branch and test it once again. Moreover, what this the expected error while spawning a jupyterlab server using empty groups?

@bdattoma User will not be able to access the jupyterhub

**Comment by prudhvi-165** [comment]
Jenkins Link : https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/936/

**Comment by bdattoma** [comment]
> > @prudhvi-165 please rebase the branch and test it once again. Moreover, what this the expected error while spawning a jupyterlab server using empty groups?
> 
> @bdattoma User will not be able to access the jupyterhub

yes, but which error page or error message will the user face?

**Review by bdattoma — CHANGES_REQUESTED** [review]

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 8]
```diff
@@ -14,6 +13,16 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
     ...       ODS-1164
     Verify Default Access Groups Settings
     Verify User Can Spawn A Notebook
+    [Teardown]  End Web Test
+
+Verify Empty Group Doesnt Allow Users To Spawn Notebooks
+    [Documentation]   Verify that User is unable to Access Jupyterhub after setting Access Groups in rhods-groups-config to Empty
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:19]: `Verifies`

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 9]
```diff
@@ -24,3 +33,23 @@ Verify User Can Spawn A Notebook
     Wait For RHODS Dashboard To Load
     Launch JupyterHub Spawner From Dashboard
     Spawn Notebook With Arguments    image=s2i-minimal-notebook    size=Default
+
+Verify User Is Unable To Spawn Notebook
+    [Documentation]    Verifies User is unable to access notebooks in jupyterhub
+    Open Browser  ${ODH_DASHBOARD_URL}  browser=${BROWSER.NAME}  options=${BROWSER.OPTIONS}
+    Login To RHODS Dashboard    ${TEST_USER.USERNAME}    ${TEST_USER.PASSWORD}    ${TEST_USER.AUTH_TYPE}
+    Wait For RHODS Dashboard To Load
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:41]: please use `Launch Dashboard` kw

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 10]
```diff
@@ -24,3 +33,23 @@ Verify User Can Spawn A Notebook
...
+    Menu.Navigate To Page    Applications    Enabled
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:42]: not needed becase the when Dashboard is launched, "Enabled" is the first page shown

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 11]
```diff
@@ -24,3 +33,23 @@ Verify User Can Spawn A Notebook
...
+    Launch JupyterHub From RHODS Dashboard Link
+    Login To Jupyterhub  ${TEST_USER.USERNAME}    ${TEST_USER.PASSWORD}    ${TEST_USER.AUTH_TYPE}
+    ${authorization_required} =  Is Service Account Authorization Required
+    Run Keyword If  ${authorization_required}  Authorize jupyterhub service account
+    Fix Spawner Status
+    Wait Until Page Contains Element  xpath://span[@id='jupyterhub-logo']
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:48]: please, use `Launch JupyterHub Spawner From Dashboard`

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 12]
```diff
@@ -24,3 +33,23 @@ Verify User Can Spawn A Notebook
...
+    Run Keyword And Expect Error    *    Wait Until JupyterHub Spawner Is Ready
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:49]: still not sure about this. It should be complemented by a check on the error the user is expected to see in the page.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 13]
```diff
@@ -15,6 +15,15 @@ Verify Default Access Groups Settings And JupyterLab Notebook Access
     Verify Default Access Groups Settings
     Verify User Can Spawn A Notebook
 
+Verify Empty Group Doesnt Allow Users To Spawn Notebooks
+    [Documentation]   Verify that User is unable to Access Jupyterhub after setting Access Groups in rhods-groups-config to Empty
+    [Tags]    Sanity
+    ...       Tier1
+    ...       ODS-572
+    Apply Access Groups Settings    admins_group=    users_group=    groups_modified_flag=true
+    Run Keyword And Expect Error  *  Verify User Can Spawn A Notebook
```
[the remark this answers is not in the document]
Comment (bdattoma) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: >> you can reproduce it manually in the cluster and see what error you should expect to receive
have you tried this?

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 11 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:48]: `Launch JupyterHub Spawner From Dashboard` contains  `Wait Until JupyterHub Spawner Is Ready` but since the user does not have required access , `Wait Until JupyterHub Spawner Is Ready` fails

Comment (bdattoma) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:48]: as expected then. That's why I'm saying to check which error is displayed..based on that, you could choose better the keywords to use

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 13 above)
[the remark this answers is not in the document]
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated using `User Is Not Allowed`

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 8 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 9 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 10 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 11 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated using `User Is Not Allowed`

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot (hunk 12 above)
Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated

File: tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot [hunk 14]
```diff
@@ -24,3 +33,18 @@ Verify User Can Spawn A Notebook
...
     Spawn Notebook With Arguments    image=s2i-minimal-notebook    size=Default
+
+Verify User Is Unable To Spawn Notebook
+    [Documentation]    Verifies User is unable to access notebooks in jupyterhub
+    Launch Dashboard    ${TEST_USER.USERNAME}  ${TEST_USER.PASSWORD}  ${TEST_USER.AUTH_TYPE}  ${ODH_DASHBOARD_URL}  ${BROWSER.NAME}  ${BROWSER.OPTIONS}
+    Launch JupyterHub From RHODS Dashboard Link
+    Login To Jupyterhub  ${TEST_USER.USERNAME}    ${TEST_USER.PASSWORD}    ${TEST_USER.AUTH_TYPE}
+    ${authorization_required} =  Is Service Account Authorization Required
+    Run Keyword If  ${authorization_required}  Authorize jupyterhub service account
+    User Is Not Allowed
```
Comment (bdattoma) [inline tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot:44]: these lines could we replaced by 2 lines only:
```
Launch JupyterHub From RHODS Dashboard Link
Login Verify Access Level  ${TEST_USER.USERNAME}
    ...                                   ${TEST_USER.PASSWORD}    ${TEST_USER.AUTH_TYPE}    none
```

Comment (prudhvi-165) [reply tests/Tests/100__deploy/110__auth_providers_and_rbac/110__rbac.robot]: Updated.

**Review by bdattoma — COMMENTED** [review]
lgtm. Please test it on jenkins

**Comment by prudhvi-165** [comment]
> lgtm. Please test it on jenkins

Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/953/

**Review by bdattoma — DISMISSED** [review]

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot [hunk 15]
```diff
@@ -99,3 +100,12 @@ Anaconda Suite Setup
   [Documentation]  Setup for ACE test suite
   Set Library Search Order  SeleniumLibrary
   RHOSi Setup
+
+Verify Anaconda In Kfdef
```
Comment (jgarciao) [inline tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:104]: This change is unrelated to this PR

**Review by jgarciao — CHANGES_REQUESTED** [review]

**Comment by prudhvi-165** [comment]
Jenkins Link: https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/976/ 
https://opendatascience-jenkins-csb-rhods.apps.ocp-c1.prod.psi.redhat.com/job/rhods-ci-pr-test/975/

File: tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot (hunk 15 above)
Comment (prudhvi-165) [reply tests/Tests/600__ai_apps/600__anaconda_commercial_edition/600__anaconda_commercial_edition.robot:104]: Updated

**Review by pablofelix — APPROVED** [review]
lgtm

**Review by jgarciao — APPROVED** [review]
lgtm
